Improve SSL context support using BouncyCastlePemReader in native images

BC provider is instantiated in `BouncyCastlePemReader` even when it is registered as Security provider. When using JVM it doesn't make a difference whether Netty instantiates the provider itself or take the provider instance from `java.security.Security#getProvider`. However when debugging https://github.com/quarkusio/quarkus/issues/46279 I have mentioned that provider instantiated during at the build time has more services then the one created during the first `BouncyCastlePemReader` usage. Registering every required BC class for reflection can be difficult task. Would it be possible to try `Security.getProvider("BC")` (`Security.getProvider("BCFIPS")` respectively) before instantiating classes using reflection API? Alternatively maybe allow users to specify BC provider themselves?

I'd appreciate any change that would allow me to use official Netty API rather than working around my issues. Thank you

### Details

I am using `4.1.118.Final`.

### Steps to reproduce

Clone `git@github.com:michalvavrik/netty-ssl-context-bouncycastle-native-issue-reproducer.git`, run `mvn clean verify -Dnative`. My `SslContextBuilder.forClient().keyManager(certificate, privateKey).build()` works with JVM:

```
2025-02-16 15:28:15,409 DEBUG [io.net.han.ssl.BouncyCastlePemReader] (executor-thread-1) Bouncy Castle provider available
2025-02-16 15:28:15,421 DEBUG [io.net.han.ssl.BouncyCastlePemReader] (executor-thread-1) Parsed PEM object of type org.bouncycastle.openssl.PEMKeyPair and assume key is not encrypted
2025-02-16 15:28:15,490 DEBUG [io.net.han.ssl.OpenSsl] (executor-thread-1) netty-tcnative not in the classpath; OpenSslEngine will be unavailable.
2025-02-16 15:28:15,510 DEBUG [io.net.han.ssl.JdkSslContext] (executor-thread-1) Default protocols (JDK): [TLSv1.3, TLSv1.2] 
2025-02-16 15:28:15,510 DEBUG [io.net.han.ssl.JdkSslContext] (executor-thread-1) Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384]
SSL context for certificate java.io.ByteArrayInputStream@7e2e2176 is io.netty.handler.ssl.JdkSslClientContext@4b6addbd

2025-02-16 15:28:15,634 DEBUG [io.net.han.ssl.BouncyCastlePemReader] (executor-thread-1) Parsed PEM object of type org.bouncycastle.asn1.pkcs.PrivateKeyInfo and assume key is not encrypted
```
 but failing when using the native image:
```
2025-02-16 15:29:56,391 DEBUG [io.net.han.ssl.BouncyCastlePemReader] (executor-thread-1) Bouncy Castle provider available
2025-02-16 15:29:56,391 DEBUG [io.net.han.ssl.BouncyCastlePemReader] (executor-thread-1) Parsed PEM object of type org.bouncycastle.openssl.PEMKeyPair and assume key is not encrypted
2025-02-16 15:29:56,391 DEBUG [io.net.han.ssl.BouncyCastlePemReader] (executor-thread-1) Unable to extract private key: org.bouncycastle.openssl.PEMException: unable to convert key pair: no such algorithm: EC for provider BC
	at org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter.getKeyPair(Unknown Source)
	at io.netty.handler.ssl.BouncyCastlePemReader.getPrivateKey(BouncyCastlePemReader.java:177)
	at io.netty.handler.ssl.BouncyCastlePemReader.getPrivateKey(BouncyCastlePemReader.java:124)
	at io.netty.handler.ssl.SslContext.toPrivateKey(SslContext.java:1204)
	at io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:421)
	at io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:348)
	at org.acme.BouncyCastleEndpoint.pkcs1(BouncyCastleEndpoint.java:39)
	at org.acme.BouncyCastleEndpoint$quarkusrestinvoker$pkcs1_b8002ac5f80c85fa5587227252617383662986bc.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:639)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@21.0.6/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.6/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:896)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:872)
Caused by: java.security.NoSuchAlgorithmException: no such algorithm: EC for provider BC
	at java.base@21.0.6/sun.security.jca.GetInstance.getService(GetInstance.java:101)
	at java.base@21.0.6/sun.security.jca.GetInstance.getInstance(GetInstance.java:218)
	at java.base@21.0.6/java.security.KeyFactory.getInstance(KeyFactory.java:266)
	at org.bouncycastle.jcajce.util.ProviderJcaJceHelper.createKeyFactory(Unknown Source)
	at org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter.getKeyFactory(Unknown Source)
	... 23 more
```

### Workaround

I have tried using `BC` provider registered as `Security` provider and it worked, but I don't want this to keep in place (Netty can easily change implementation):

```
@com.oracle.svm.core.annotate.TargetClass(className = "io.netty.handler.ssl.BouncyCastlePemReader", onlyWith = NettySslBountyCastleSupportRequired.class)
final class Target_io_netty_handler_ssl_BouncyCastlePemReader {
    @Alias
    private static volatile Provider bcProvider;
    @Alias
    private static volatile boolean attemptedLoading;
    @Alias
    private static volatile Throwable unavailabilityCause;

    @com.oracle.svm.core.annotate.Substitute
    public static boolean isAvailable() {
        if (!attemptedLoading) {
            bcProvider = Security.getProvider(SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_NAME);
            if (bcProvider == null) {
                bcProvider = Security.getProvider(SecurityProviderUtils.BOUNCYCASTLE_FIPS_PROVIDER_NAME);
            }
            if (bcProvider == null) {
                tryLoading();
            } else {
                attemptedLoading = true;
            }
        }
        return unavailabilityCause == null;
    }

    @Alias
    private static void tryLoading() {

    }
}
```

--- a/handler/src/main/java/io/netty/handler/ssl/BouncyCastlePemReader.java
+++ b/handler/src/main/java/io/netty/handler/ssl/BouncyCastlePemReader.java
@@ -42,9 +42,12 @@ import java.security.AccessController;
 import java.security.PrivateKey;
 import java.security.PrivilegedAction;
 import java.security.Provider;
+import java.security.Security;
 
 final class BouncyCastlePemReader {
+    private static final String BC_PROVIDER_NAME = "BC";
     private static final String BC_PROVIDER = "org.bouncycastle.jce.provider.BouncyCastleProvider";
+    private static final String BC_FIPS_PROVIDER_NAME = "BCFIPS";
     private static final String BC_FIPS_PROVIDER = "org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider";
     private static final String BC_PEMPARSER = "org.bouncycastle.openssl.PEMParser";
     private static final InternalLogger logger = InternalLoggerFactory.getInstance(BouncyCastlePemReader.class);
@@ -77,21 +80,27 @@ final class BouncyCastlePemReader {
             public Void run() {
                 try {
                     ClassLoader classLoader = getClass().getClassLoader();
-                    // Check for bcprov-jdk15on or bc-fips:
-                    Class<Provider> bcProviderClass;
-                    try {
-                        bcProviderClass = (Class<Provider>) Class.forName(BC_PROVIDER, true, classLoader);
-                    } catch (ClassNotFoundException e) {
-                        try {
-                            bcProviderClass = (Class<Provider>) Class.forName(BC_FIPS_PROVIDER, true, classLoader);
-                        } catch (ClassNotFoundException ex) {
-                            ThrowableUtil.addSuppressed(e, ex);
-                            throw e;
-                        }
-                    }
                     // Check for bcpkix-jdk15on:
                     Class.forName(BC_PEMPARSER, true, classLoader);
-                    bcProvider = bcProviderClass.getConstructor().newInstance();
+                    // Check for bcprov-jdk15on or bc-fips:
+                    bcProvider = Security.getProvider(BC_PROVIDER_NAME);
+                    if (bcProvider == null) {
+                        bcProvider = Security.getProvider(BC_FIPS_PROVIDER_NAME);
+                    }
+                    if (bcProvider == null) {
+                        Class<Provider> bcProviderClass;
+                        try {
+                            bcProviderClass = (Class<Provider>) Class.forName(BC_PROVIDER, true, classLoader);
+                        } catch (ClassNotFoundException e) {
+                            try {
+                                bcProviderClass = (Class<Provider>) Class.forName(BC_FIPS_PROVIDER, true, classLoader);
+                            } catch (ClassNotFoundException ex) {
+                                ThrowableUtil.addSuppressed(e, ex);
+                                throw e;
+                            }
+                        }
+                        bcProvider = bcProviderClass.getConstructor().newInstance();
+                    }
                     logger.debug("Bouncy Castle provider available");
                     attemptedLoading = true;
                 } catch (Throwable e) {
@@ -102,6 +111,19 @@ final class BouncyCastlePemReader {
                 return null;
             }
         });
+    }
+
+    /**
+     * Allows to test {@link #attemptedLoading} under different conditions.
+     *
+     * @return previous {@link #bcProvider} value
+     */
+    static Provider resetBcProvider() {
+        Provider previousProvider = bcProvider;
+        bcProvider = null;
+        attemptedLoading = false;
+        unavailabilityCause = null;
+        return previousProvider;
     }
 
     /**

--- a/handler/src/test/java/io/netty/handler/ssl/BouncyCastlePemReaderTest.java
+++ b/handler/src/test/java/io/netty/handler/ssl/BouncyCastlePemReaderTest.java
@@ -1,0 +1,72 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.parallel.Isolated;
+
+import java.security.Provider;
+import java.security.Security;
+
+import static java.security.Security.addProvider;
+import static java.security.Security.removeProvider;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@Isolated("Provider value stored in the BouncyCastlePemReader#bcProvider field is also used by other tests")
+public class BouncyCastlePemReaderTest {
+
+    @Test
+    public void testBouncyCastleProviderLoaded() {
+        // tests org.bouncycastle.jce.provider.BouncyCastleProvider is detected as available
+        // because provider with matching name is present in 'java.security.Security'
+
+        assertTrue(BouncyCastlePemReader.isAvailable());
+        Provider bcProvider = BouncyCastlePemReader.resetBcProvider();
+        assertNotNull(bcProvider);
+
+        Provider bouncyCastleProvider = new BouncyCastleProvider();
+        Security.addProvider(bouncyCastleProvider);
+        assertTrue(BouncyCastlePemReader.isAvailable());
+        bcProvider = BouncyCastlePemReader.resetBcProvider();
+        assertSame(bouncyCastleProvider, bcProvider);
+        Security.removeProvider(bouncyCastleProvider.getName());
+    }
+
+    @Test
+    public void testBouncyCastleFipsProviderLoaded() {
+        // tests org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider is detected as available
+        // because provider with matching name is present in 'java.security.Security'
+
+        assertTrue(BouncyCastlePemReader.isAvailable());
+        Provider bcProvider = BouncyCastlePemReader.resetBcProvider();
+        assertInstanceOf(BouncyCastleProvider.class, bcProvider);
+
+        // we don't expect to have both BC and BCFIPS available, but BouncyCastleProvider is on the classpath
+        // hence we need to add a fake BouncyCastleFipsProvider provider
+        Provider fakeBouncyCastleFipsProvider = new Provider("BCFIPS", 1.000205,
+                "BouncyCastle Security Provider (FIPS edition) v1.0.2.5") { };
+        Security.addProvider(fakeBouncyCastleFipsProvider);
+        assertTrue(BouncyCastlePemReader.isAvailable());
+        bcProvider = BouncyCastlePemReader.resetBcProvider();
+        assertSame(fakeBouncyCastleFipsProvider, bcProvider);
+        Security.removeProvider(fakeBouncyCastleFipsProvider.getName());
+    }
+
+}

